dvc/remote/base.py

```diff
@@ -942,23 +965,33 @@ def _checkout(
         force=False,
         progress_callback=None,
         relink=False,
+        filter_info=None,
     ):
         if not self.is_dir_checksum(checksum):
             return self._checkout_file(
                 path_info, checksum, force, progress_callback=progress_callback
             )
         return self._checkout_dir(
-            path_info, checksum, force, progress_callback, relink
+            path_info, checksum, force, progress_callback, relink, filter_info
         )
 
-    def get_files_number(self, checksum):
+    def get_files_number(self, path_info, checksum, filter_info):
+        from funcy.py3 import lfilter
+
         if not checksum:
             return 0
 
-        if self.is_dir_checksum(checksum):
+        if not self.is_dir_checksum(checksum):
+            return 1
+
+        if not filter_info:
             return len(self.get_dir_cache(checksum))
 
-        return 1
+        def _func(entry):
+            info = path_info / entry[self.PARAM_RELPATH]
+            return info.isin_or_eq(filter_info)
+
+        return len(lfilter(_func, self.get_dir_cache(checksum)))
```

Comment:
```suggestion
        return ilen(filter(_func, self.get_dir_cache(checksum)))
```
to not create an intermediate list, here is [ilen()](https://funcy.readthedocs.io/en/stable/seqs.html#ilen) :)

Reply:
And maybe drop `.isin_or_eq()` and make a `PathInfo.contains()` instead:
```python
return ilen(
    filter_info.contains(path_info / entry[self.PARAM_CHECKSUM])
    for entry in self.get_dir_cache(checksum)
)
```